Feature - Website Drafts

dragonstyle opened the pull request:
This feature improves Quarto's support for draft documents in websites. It does this a few ways:

### Improved linking behavior for draft documents

Previously, draft documents were excluded from search results, listings, and the sitemap. Now, draft documents will not appear in navigation as well (sidebar, navbars, and footers). Entries that link to a draft document will be removed / excluded. In addition, if a page on the website links to a draft document, the link will be omittted (leaving the contents of the link without the hyperlink itself). 

These changes add up to mean that now when you mark a document as a draft, other pages will not link to it, so it will be as if it isn't yet a part of the site.

### Draft Modes

You can use the `draft-mode` option to control the behavior of draft documents in the rendered website. Use the following to control the behavior:

- `visible` - the draft will visible and fully available
- `unlinked` - the draft will be rendered, but will not appear in navigation, search, or listings.
- `gone` - the draft will have no content and will not be linked to.

When the `gone` mode is selected, each draft document will still produce a file when the site is rendered, but the document will be empty.

### Website Project Drafts

You can provide a list of prroject relative paths that will be considered a draft in your project configuration using the `drafts` option with an array of paths. Input listed in the option will be considered drafts and treated accodring to the draft mode that is selected.

Note that you can use `metadata-files` or `profile` to externalize the list of drafts (for example, if you wish to programmatically generate a list of draft documents).

### Complete Example

**_quarto.yml**
```yaml
project:
  type: website

website:
  title: "Cool Website."
  navbar:
    left:
      - stuff/item1.qmd
      - stuff/item2.qmd
      - stuff/item3.qmd
      - listing.qmd
      - text: Another One
  sidebar:
    contents: stuff
  page-footer: 
    center: 
      - stuff/item1.qmd
      - stuff/item3.qmd
  drafts:
    - stuff/item3.qmd
  draft-mode: visible      

format:
  html:
    theme: cosmo
    css: styles.css
    toc: true
  
metadata-files:
  - drafts.yml
```

### Complete Example with External Draft List

**_quarto.yml**
```yaml
project:
  type: website

website:
  title: "Cool Website."
  navbar:
    left:
      - stuff/item1.qmd
      - stuff/item2.qmd
      - stuff/item3.qmd
      - listing.qmd
      - text: Another One
  sidebar:
    contents: stuff
  page-footer: 
    center: 
      - stuff/item1.qmd
      - stuff/item3.qmd

format:
  html:
    theme: cosmo
    css: styles.css
    toc: true
  
metadata-files:
  - drafts.yml
```

**drafts.yml**
```yaml
website:
  drafts:
    - stuff/item3.qmd
  draft-mode: gone
 ```

Fixes #2781

dragonstyle committed:
Website Drafts - inline document links

This changes the system behavior when forming links to inputs whose metadata directly contains ‘draft: true’. Previously, ‘draft’ documents would be excluded from search, the sitemap, and listings, but would still appear in navigation and in links in content.

This resolves the link in content, removing the link if the document is marked ‘draft’ (and leaving the text, as is).

dragonstyle committed:
Website Drafts - navigation

When draft documents are encountered when rendering navigation, just drop them.

- navbar
- sidebar (explicit, auto)
- page-footer

This will be extended to support draft modes that control behavior, allowing preview to show draft content while published sites do not show draft content.

cscheid commented on src/project/project-index.ts:
```diff
@@ -107,6 +109,9 @@ export async function readBaseInputIndex(
         | undefined,
     markdown,
     formats,
+    draft: (firstFormat?.metadata?.[kDraft] || markdown.yaml?.[kDraft]) as
```
> Note that this only applies to draft when set directly in document metadata (not via _metadata.yml or other merging metadata).

Is this the place where we would in the future use project-wide information to determine draft status?

dragonstyle committed:
Website Drafts - empty draft docs

Empties draft documents, so that when they are published they are devoid of content. This will be controlled by a draft mode that either allows the content through or removes the content (currently `draft-mode`, but that is a work in progress)

dragonstyle committed:
Website Drafts - centralize input draft check

Update search, listings, sitemap, etc.. to properly use the shared way to determine whether an input is a draft.

This will set the stage for adding a project level option for drafts.

dragonstyle committed:
Website Drafts - project options

Add project options to website schema for controlling drafts. Options are:

`visible` - the draft will visible and fully available
`unlinked ` - the draft will be rendered, but will not appear in navigation, search, or listings. (default)
`gone` - the draft will have no content and will not be linked to.

Provides a list of draft documents within the project. This will be used to be injected into website projects via environment or metadata-files and will allow centralized control of whether a file is a draft or not.

Example:

```yaml
project:
  type: website

website:
  title: "Cool Website, Buddy."
  navbar:
    left:
      - stuff/item1.qmd
      - stuff/item2.qmd
      - stuff/item3.qmd
      - listing.qmd
      - text: Another One
  sidebar:
    logo: salute.gif
    align: center
    contents: stuff
    tools:
      - icon: twitter
        href: https://twitter.com
    style: floating
  page-footer:
    center:
      - stuff/item1.qmd
      - stuff/item3.qmd
  draft-mode: unlinked
  drafts:
    - stuff/item3.qmd

format:
  html:
    theme: cosmo
    css: styles.css
    toc: true

```

dragonstyle committed:
Website Drafts - projectype input filter

Allow project type types to filter the input target, which will be used by website project to decorate drafts.

dragonstyle committed:
Website Drafts - Website proj filter

The website project type will now filter input targets to include the draft status if explicitly provided in the project file.

dragonstyle committed:
Website Draft - filter params

Forward the drafts and draft-mode to the draft filter so it can determine whether to mark the output as a draft and clear it.

dragonstyle committed:
Website Draft - ejs respect draftmode

EJS templates for emitting navigation HTML will now respect draftmode - only dropping draft items when the draft mode is not `visible`

dragonstyle committed:
Website Drafts - centralize project drafts

- Centralize the check for project drafts.
- Use that check when resolving document draft links.
- Update nav resolution to use it as well.

dragonstyle committed:
Website Draft - listings fix

Since listings are actually generated pre-render, the input index isn’t yet populated. So this needs to more explicitly check the project render list directly.

ick

Check draft mode listing

dragonstyle committed:
Website Drafts - Testing

Add test website which can be rendered as a part of testing. Site supports differing draft modes:

- profile `drafts` will include the post `draft-post` in the project draft list.
- `drafts-gone` and `drafts-visible` will change `draft-mode`

Use like:

`quarto render —profile drafts,drafts-visible`
`quarto render —profile drafts,drafts-gone`

test site updates

Website Drafts - baseline testing

A test that will verify render behavior of the various draft modes in the proper locations.

dragonstyle committed:
Website Drafts - Sitemap, search draftmode support

Add support for draftmodes in search and sitemap generation (particularly visible drafts).

dragonstyle committed:
Website Drafts - Improved path handling

Ensure that we’re passing expected looking paths to resolve input targets (on windows in particular).

dragonstyle committed:
Website Drafts -  centralize draft visiblity check

dragonstyle committed:
Project Init - reload context pre-render

When running the pre-render scripts, we have already initialized the project context, which is passed through the rest of the system. The issue with this is that if the pre-render script modifies anything that would appear in the ProjectContext, those changes aren’t reflected downstream.

The simplest example would be generating a file to be included via `metadata-files` - the generated file isn’t present before the pre-render script runs (when the ProjectContext is first loaded), so it doesn’t affect the downstream render as it shows up too late.

This change will load the ProjectContext as normal, but then re-initialize it after the post render script has executed.

Note that certain project data will not be mutable in pre-render, as it is data that is essentially passed to pre-render:

- output-dir
- render list
- projDir
- projType

This is very much included here for review and careful contemplation.

dragonstyle committed:
Website Drafts - Preprocessor generation test

dragonstyle committed:
Website Drafts - draft notice

Display a draft notice for visible drafts.

dragonstyle committed:
Website Drafts - adorn drafts in preview

Switch draft mode to visible when previewing.

correct preview

dragonstyle committed:
Project Init - error on invalid mutation

Some properties may not be mutated by the pre-render script. Error if those properties become mutated.

dragonstyle committed:
Project Init - Tests

Simple tests of project rendering. Validates that a mutation error is thrown properly, and validates that a render will pick up generated QMD files.

dragonstyle committed:
Project Init - Reload render config

After the pre render script has been executed, refresh the project render configuration (including things like the render list and supplements). This should allow pre-render scripts to do things like generate QMD files which should be included in the pre-render step.

dragonstyle committed:
Website Draft - default to gone

update test
